### PR TITLE
Claims pallet

### PR DESCRIPTION
**Pull request by green-jay**

### CODE REVIEW

File: pallets/claims/src/lib.rs
Comment (enthusiastmartin) [inline pallets/claims/src/lib.rs:106]: Nice. Would be cool to come up with a test that tests it somehow.

File: pallets/claims/src/tests.rs
Comment (enthusiastmartin) [inline pallets/claims/src/tests.rs:20]: In the tests - we need to verify the balance after the claim as well.

**Review by enthusiastmartin — CHANGES_REQUESTED** [review]

File: pallets/claims/src/lib.rs
Comment (enthusiastmartin) [inline pallets/claims/src/lib.rs:111]: actually, you can move the event and OK inside the with transaction and use that to return the result directly.

File: pallets/claims/src/tests.rs
Comment (enthusiastmartin) [inline pallets/claims/src/tests.rs:23]: I would also verify that the balance has not changed in any way if error occurs.

File: pallets/claims/src/benchmarking.rs
Comment (enthusiastmartin) [inline pallets/claims/src/benchmarking.rs:19]: Benchmark can have verify section where you can check if the claim has been successful. Similar to test - i would check the balance here to make sure that it is correct.

File: pallets/claims/src/lib.rs
Comment (enthusiastmartin) [inline pallets/claims/src/lib.rs:75]: This error does not have tests

File: pallets/claims/Cargo.toml
Comment (enthusiastmartin) [inline pallets/claims/Cargo.toml:41]: We should use the same version - 0.3.3-dev. I believe the reason why it is not found when you try to use it - is that it is not patched in the Cargo file in the top level repo dir. If you add the orml-utilies to patch section of that cargo file with the same revision as the other orml crates - you will be able to use 0.3.3-dev version.

File: pallets/claims/src/lib.rs
Comment (enthusiastmartin) [inline pallets/claims/src/lib.rs:38]: I was just thinking that there is no need to use multi currency in this pallet. Claims only deal with native/core asset. So we could just use the Currency from the frame support. And then use balances pallet to handle native currency. 

just for info: multi currency depends on the native currency "balances" anyway. And if you deposit core asset like you do here in this pallet - the multi currency just calls the balances deposit. So we can get rid of the multi currency layer and call balances directly.

Comment (green-jay) [reply pallets/claims/src/lib.rs:38]: I was thinking similarly and originally used the currency type. But when talking about DEX, it just seemed more natural and correct to use the multicurrency type and explicitly define that it is the core asset we want to deposit even though it is not required here. It is also aligned with other pallets. Shall we modify it regardless?

Comment (enthusiastmartin) [reply pallets/claims/src/lib.rs:38]: Yes, this makes sense too. But i wonder if using multi currency support here would require also adding additioan tests to make sure that we do not deposit to any other currency ( might sound farfetched?!, not sure). I will discuss this with @jak-pan and we'll see.

File: pallets/claims/src/lib.rs
Comment (jak-pan) [inline pallets/claims/src/lib.rs:47]: Pls rename this to ClaimData or something more indicative of what it is.

File: pallets/claims/src/benchmarking.rs
Comment (jak-pan) [inline pallets/claims/src/benchmarking.rs:18]: Try to not use HDX as we might want to open-source this as a generic pallet.

File: pallets/claims/src/lib.rs
Comment (green-jay) [reply pallets/claims/src/lib.rs:47]: IMHO, the "Data" does not provide more clarity but I can refactor the storage name and remove HDX from it. "Claims" as the name of storage is used for multiple projects including Polkadot, so I would stick with that.

File: pallets/claims/src/lib.rs
Comment (jak-pan) [reply pallets/claims/src/lib.rs:38]: Your call. 

We can simplify to be quick and safe. 
If we actually do this in a generic way. We can probably pr this to orml? It would need more testing tho.

First approach is more general. Second approach is more advanced and we're limiting users so that they need to use orml. 

Q to think about: Can this be a configurable trait? Can we do without id if no asset_id is present and have it if it is present?

File: pallets/claims/src/lib.rs
Comment (jak-pan) [inline pallets/claims/src/lib.rs:90]: https://github.com/paritytech/substrate/blob/171ad2322d7561934b82f0b7e41c58d1710f8c24/primitives/runtime/src/traits.rs#L777

From: Alexander Popiak
Answering the questions about transaction propagation:
Txs that return error when passed to validate_transaction here are not propagated.

Comment (jak-pan) [reply pallets/claims/src/lib.rs:90]: cc @enthusiastmartin

File: runtime/src/lib.rs
Comment (jak-pan) [inline runtime/src/lib.rs:350]: ClaimMessagePrefix

File: pallets/claims/src/mock.rs
Comment (jak-pan) [inline pallets/claims/src/mock.rs:68]: "HDX tokens to wallet: "?

Comment (green-jay) [reply pallets/claims/src/mock.rs:68]: The whitespace could do more harm than good. I agree with change to HDX.

File: pallets/claims/src/lib.rs
Comment (green-jay) [reply pallets/claims/src/lib.rs:39]: As a fan of the Occam's razor, I would choose quick and safe. I should be ready with a commit tomorrow.

File: pallets/claims/src/lib.rs
Comment (enthusiastmartin) [inline pallets/claims/src/lib.rs]: I would suggest to use like this :

type CurrencyBalance: Into<BalanceOf<Self>> + From<Balance>;

and order it so Currency and CurrencyBalance is together , and make a comment that this is the currency used in Currency and needs to support conversion to Balance (as we use it directly for the actual data )

File: pallets/claims/src/lib.rs
Comment (green-jay) [reply pallets/claims/src/lib.rs:91]: can be resolved or still a concern?

Comment (green-jay) [reply pallets/claims/src/lib.rs:91]: @jak-pan @enthusiastmartin

File: pallets/claims/src/mock.rs
Comment (lumir-mrkva) [reply pallets/claims/src/mock.rs:57]: any specific reason not to add space after : ? it looks like typo when you are signing

File: pallets/claims/src/mock.rs
Comment (lumir-mrkva) [inline pallets/claims/src/mock.rs:93]: please add hex representation and attach private key of the test address

**Review by lumir-mrkva — APPROVED** [review]

File: node/src/chain_spec.rs
Comment (Roznovjak) [reply node/src/chain_spec.rs:478]: I am not sure what you mean. Claim balances are tied to eth addresses, not to our accounts. Do you mean to create claim balances and sign messages for all `endowed_accounts`?

File: pallets/claims/src/mock.rs
Comment (green-jay) [reply pallets/claims/src/mock.rs:58]: It is a string that user will copy&paste. If e.g. was copied as two parts, prefix and later address, the whitespace could cause some confusion. It is imho more visible and clear not to have it than have it. Comes down to personal preference I guess. Don't mind either way

File: pallets/claims/src/mock.rs
Comment (green-jay) [reply pallets/claims/src/mock.rs]: done

File: node/src/chain_spec.rs
Comment (lumir-mrkva) [reply node/src/chain_spec.rs:478]: ah, sorry for the confusion, you are correct. is just important to have claims pre-generated to addresses for local dev chain with known private keys to be usable

File: pallets/claims/src/lib.rs
Comment (green-jay) [reply pallets/claims/src/lib.rs:106]: txs validated by SignedExtension

File: node/src/chain_spec.rs
Comment (green-jay) [reply node/src/chain_spec.rs:478]: test seed added

File: pallets/claims/src/migration.rs
Comment (green-jay) [reply pallets/claims/src/migration.rs:5]: agreed, changed to import_initial_claims

File: pallets/claims/src/lib.rs
Comment (lumir-mrkva) [inline pallets/claims/src/lib.rs:77]: Events shouldn't be in past tense `Claimed` -> `Claim` and this particular event should contain ETH address from which was this balance claimed.

**Review by lumir-mrkva — CHANGES_REQUESTED** [review]

File: pallets/claims/src/lib.rs
Comment (lumir-mrkva) [reply pallets/claims/src/lib.rs:77]: Currently, the same information will be laso emitted in the `Endowed` event so the ETH claim address is the most useful information transmitted by this event.

File: pallets/claims/src/lib.rs
Comment (lumir-mrkva) [inline pallets/claims/src/lib.rs:43]: is there any reason for the prefix being configurable?

File: pallets/claims/src/lib.rs
Comment (enthusiastmartin) [reply pallets/claims/src/lib.rs:106]: handled by the signed extension to validate claim.

File: pallets/claims/src/lib.rs
Comment (green-jay) [reply pallets/claims/src/lib.rs:77]: In substrate recipe describing events there is the present tense so we can stick with that.
ETH address added

File: pallets/claims/src/lib.rs
Comment (green-jay) [reply pallets/claims/src/lib.rs:43]: yes, it was designed bearing in mind we might want to publish the claims as a standalone pallet

Comment (jak-pan) [reply pallets/claims/src/lib.rs:43]: It can be used by anybody else then as it's pretty generic

File: node/src/chain_spec.rs
Comment (jak-pan) [inline node/src/chain_spec.rs:485]: Add private key to comment.

File: node/src/chain_spec.rs
Comment (jak-pan) [inline node/src/chain_spec.rs:490]: Remove these before merging

File: pallets/claims/src/claims_data.rs
Comment (jak-pan) [inline pallets/claims/src/claims_data.rs:5]: We'll replace this with real data right?

File: pallets/claims/src/lib.rs
Comment (jak-pan) [inline pallets/claims/src/lib.rs:21]: Consolidate imports. Got multiple lines of the same package. You can use the ::{} syntax

Comment (jak-pan) [reply pallets/claims/src/lib.rs:21]: Also reorder pls

File: pallets/claims/src/claims_data.rs
Comment (green-jay) [reply pallets/claims/src/claims_data.rs:5]: yes, if those can be published (otherwise we could restrict file permissions on github or other solution in mind?)

File: pallets/claims/src/migration.rs
Comment (jak-pan) [inline pallets/claims/src/migration.rs:33]: We need more tests.

- negative balances
- strings
- floats
- duplicated address
- zero balance
- overflowed balance

Also we don't handle these in migration and we probably should.

File: pallets/claims/src/claims_data.rs
Comment (jak-pan) [reply pallets/claims/src/claims_data.rs]: Of course they need to be published :)

**Review by jak-pan — CHANGES_REQUESTED** [review]
Overall looks very nice just some tidying up and maybe a few more migration tests and we're good

File: node/src/chain_spec.rs
Comment (green-jay) [reply node/src/chain_spec.rs:494]: Richard just added some additional info like private keys to claim2 & claim3, remove anyway?

File: pallets/claims/src/migration.rs
Comment (Roznovjak) [reply pallets/claims/src/migration.rs:33]: I think it's too late to do any tests at runtime (in this case, since data migration is a one-time operation handled by responsible devs/chain maintainers).

File: node/src/chain_spec.rs
Comment (Roznovjak) [reply node/src/chain_spec.rs:494]: The question is What data do we want to use for claims in local testnets?

Comment (jak-pan) [reply node/src/chain_spec.rs:494]: Ok if we have priv keys we're good. But add some reasonable amounts to them then. 
0.001, 1, and 1000 * 10^12 tokens should be good?

Comment (Roznovjak) [reply node/src/chain_spec.rs:494]: Is 1 equal to 10^12 ?

File: pallets/claims/src/migration.rs
Comment (jak-pan) [reply pallets/claims/src/migration.rs:33]: How do we generate the data into .rs file? @lumir-mrkva

File: node/src/chain_spec.rs
Comment (jak-pan) [reply node/src/chain_spec.rs:494]: Yes.

Here:
https://github.com/galacticcouncil/HydraDX-node/blob/496178440b9ed980d58ab07cecfc3c0432a7e877/runtime/src/constants.rs#L4

And here:
https://github.com/galacticcouncil/HydraDX-node/blob/fadfe28ff04abf504caa0adea475d2a3e04c82d8/node/src/chain_spec.rs#L87

File: pallets/claims/src/migration.rs
Comment (lumir-mrkva) [reply pallets/claims/src/migration.rs:33]: i agree with @Roznovjak, migration doesn't need to be tested so thoroughly, besides we can test it with real data that we'll do real migration with

Comment (jak-pan) [reply pallets/claims/src/migration.rs:33]: OK but then we need to definitely test the source script that generates it.

Comment (Roznovjak) [reply pallets/claims/src/migration.rs:33]: We can create public subrepo for the script, or add the script to the `claims pallet` folder (or some `script` folder?).

Comment (green-jay) [reply pallets/claims/src/migration.rs:33]: Will be resolved separately [here](https://github.com/galacticcouncil/xhdx) I suppose?

File: node/src/chain_spec.rs
Comment (jak-pan) [inline node/src/chain_spec.rs:487]: Will this actually work? (upper/lowercase)

Comment (Roznovjak) [reply node/src/chain_spec.rs:487]: Tests pass. AFAIK same wallets use capitalization to compute a checksum and validate an address. Hex conversion algorithms should be clever enough to handle both lower and uppercase input. ('a' and 'A' are the same in hex).

Comment (jak-pan) [reply node/src/chain_spec.rs:487]: We should probably test this with metamask as I know that there were some problems with this and also we're lowercasing addresses in UI. I'll do that when I have time.

Comment (Roznovjak) [reply node/src/chain_spec.rs:487]: I've generated all signatures using mycrypto + metamask. It's also easy to convert 3 public keys to lowercase and update the source code.

**Review by lumir-mrkva — APPROVED** [review]

File: node/src/chain_spec.rs
Comment (jak-pan) [reply node/src/chain_spec.rs:487]: It looks like it works anyway so all good.

File: pallets/claims/src/lib.rs
Comment (lumir-mrkva) [inline pallets/claims/src/lib.rs:49]: consider using common `PalletVersion` storage for this instead

Comment (lumir-mrkva) [reply pallets/claims/src/lib.rs:49]: here is the change https://github.com/galacticcouncil/HydraDX-node/pull/132

**Review by jak-pan — APPROVED** [review]
LGTM. Tested. Let's go 🤘

**Pull request merged**
